feature: Command line arguments - version and platform

haadcode opened the pull request:
This PR will:

Add possibility to set the wanted go-ipfs version in package.json.
Add possibility to specify the version to download.
Add possibility to specify the platform to download.
Add possibility to specify the arch to download.
Add possibility to specify the download url/domain.
New public API.
Add simple stdout output for download progress.
Catch errors and add output them to stderr.
Add stdout output for user feedback.
Add more tests.

haadcode commented:
@diasdavid @dignifiedquire can I get a CR for this? This will add the ability to specify version and platform and the module becomes 200% more useful. This will also make it work on CI to build for any platform (eg. together with Electron), ie. I can include Windows version of go-ipfs on Linux machine.

It's starting work quite nicely as a "go-ipfs version manager", perhaps at some point we could write one on top of this.

dignifiedquire commented on package.json:
```diff
@@ -4,7 +4,7 @@
   "main": "src/index.js",
   "scripts": {
-    "install": "node src/bin.js",
+    "install": "node src/bin.js $TARGET_VERSION $TARGET_OS",
```
This will not work on windows, but there is https://www.npmjs.com/package/cross-env to fix that

haadcode replied:
Is there any cross platform solution that doesn't require pulling in a dependency?

dignifiedquire replied:
not as far as I know, which is why that package exists in the first place

haadcode replied:
What happens on Windows with the command? Does `process.argv` get the params as strings or nothing? If latter, then I think we're fine for now, when the params are undefined, it'll download the current platform and version from the package.

If they do get passed in as strings on windows, then we need to figure out something better.

dignifiedquire replied:
it probably throws because the envvariables are not valid syntax on windows shells

dignifiedquire replied:
so this will literally stop any windows users from installing

dignifiedquire commented on src/bin.js:
```diff
@@ -2,4 +2,6 @@
 
 'use strict'
 
-require('./')()
```
why the change of permissions?

haadcode replied:
Opps, didn't need to. Will revert that change.

haadcode replied:
Changed back to 644

haadcode replied:
Oh, I can't :/

dignifiedquire commented on src/bin.js:
```diff
@@ -2,4 +2,6 @@
-require('./')()
+// First param is the target version
+// Second param is the target platform
+require('./')(process.argv[2], process.argv[3], function () {})
```
Rather than passing in an empty function, why not make the callback either optional or useful and print something like `successfully installed <version>`

haadcode replied:
Later. Right now it works fine without.

diasdavid replied:
Can we make that 'later' more explicit? With a TODO of what is desired for example.

haadcode commented:
I removed the env variables from the npm script to the actual version setup in index.js. This way it can pick up the platform and version from a) arguments passed in b) environment variables and c) defaults (package.json and go-platform).

Also cleaned up the callback so we don't need to discuss what we put there. And changed the file permission back to original. 

Everything addressed, can we get final CR and if all good, merge this?

dignifiedquire commented on src/index.js:
```diff
@@ -7,26 +7,48 @@ var path = require('path')
   // hacky hack hack to work around unpublishability
-  version = version.replace(/-[0-9]+/, '')
+  // version = version.replace(/-[0-9]+/, '')
+  // What is this^^?? @haadcode
```
Not sure we can drop this, because the version in `package.json` corresponds to go version, but we sometimes have multiple versions of this package per go-ipfs version, so this fixes that, i.e.

| go-ipfs | package.json |
| 0.4.4    | 0.4.4    |
| 0.4.4    | 0.4.4-1 | 
| 0.4.4    | 0.4.4-2 |
| 0.4.4    | 0.4.4-3 |

dignifiedquire commented on src/index.js:
```diff
@@ -1,67 +1,121 @@
+Object.assign(download, { Platforms: support.Platforms })
+Object.assign(download, { Versions: support.Versions })
+Object.assign(download, { Archs: support.Archs })
+Object.assign(download, { Download: download })
```
As you are not checking for existence, you could just

```js
download.Platforms = support.Platforms,
download.Versions = support.Versions
```

also why are these captialized, they are not constructors, so I think they should be lower case

dignifiedquire replied:
how is direct assignment less clearer than using `Object.assign`?

dignifiedquire replied:
For me it's the exact opposite, also why not just use a single call to `Object.assign`? I had to go through the whole code checking if this wasn't overriding something because that is what `Object.assign` is usually used for, merging objects with possible conflicts

dignifiedquire replied:
Though to be honest, I think this would be a much better interface, than assigning props to a function

```js
exports.download = download
exports.platforms = support.platforms
exports.versions = support.versions
exports.archs = support.archs
```

haadcode replied:
I refactored the public api. I still think the uppercased properties are nicer, so keeping them as such. We need to have the download function as our module.export in order to not break the API for users (backwards compatibility).

haadcode replied:
Take that back, that doesn't compile :)

haadcode replied:
You know what, let's take the Versions, Platforms and Archs out of the public api. They're not really what this module *needs* to do, so let's keep it the way it was. If we need such functionality in the future, we can add something like `getAllPackages()` or smth that then returns the data.

This way the code is the clearest out of all options and we don't add a need to support a larger api surface :)

haadcode replied:
Changed

haadcode commented:
Everything got refactored and updated. Added more thorough support for arguments (arch, dist url). Squashed all commits.

Ready for review!

dignifiedquire commented on Makefile:
```diff
@@ -0,0 +1,15 @@
+all: test
```
I am 👎 on adding makefiles, they are not cross platform and don't add much value compared to just `npm run-script`

haadcode replied:
I can remove the Makefile from this PR so that we don't need to bikeshed it here. Give the Makefile a try, you'll be surprised how nice it makes the dev process and how it actually documents what is done at each step.

haadcode replied:
That said, I'm highly in favor of adding Makefiles. @jbenet's been going on about them for a long time and I didn't pay attention until very recently and he's totally right about their value. They just make it so much easier to get started than decoding what's in a specific package.json file.

jbenet replied:
> I am 👎 on adding makefiles, they are not cross platform and don't add much value compared to just npm run-script

Wrong. They add an enormous amount of value for users of tools cross language. The point of `make` is that -- no matter who wrote something and what they use daily -- a `Makefile` is a language-agnostic tool, so anyone can jump in and understand what to do. I'm not suggesting you use `make` to do the heavy lifting. I'm suggesting the use of `Makefiles` in places like Orbit, or orbit-db, where people will download the repos and want to do something, but (a) don't know npm well. (b) don't know how to invoke things. (c) dont know _which_ things to invoke. `make` is a _standard_ way to do it.

also, i dont know what the current distribution of people with Windows, unix tooling, and no make program. It used to be quite small. maybe someone should compile `make` with emscripten and publish it to npm.

jbenet replied:
Also, i'm not advocating the horrendous autotools or Cmake makefiles that are so confusing and hard to figure out. I'm advocating basic things like this:
- https://github.com/jbenet/macos-ssh-proxy/blob/master/Makefile
- https://github.com/jbenet/go-random-files/blob/master/Makefile
- https://github.com/jbenet/gob-js-test/blob/master/Makefile (particularly see the deps part)

obviously none of this is perfect, and i dont use them in all my repos, but i've learned that it's a great way to signal "what this thing should do" and "how to use this thing" in a language everyone can understand.

haadcode replied:
👍  to all what @jbenet said: keep 'em simple, use them to document *how* to get a repo up an running (*which* things to run is the most important description here), not for heavy lifting (npm is fine for that in JS, other langs have their own tools).

@jbenet also: cool idea to add command help to `make all`.

haadcode replied:
What I've found hugely valuable with Makefiles is with the "immutability". `make clean` makes sure to reset the repo to its original state and testing becomes much more reliable when `node_modules/` gets wiped and re-installed from clean slate. This was the reason I started adding them to my repos. They give a really simple and clean set of instructions to run on a repo to make sure everything is working: `make clean && make` and all the expected things should've been run & built (dist builds, tests, installing deps, etc).

I would recommend the following approach: try it out, add it to the repos you work on and see if it is valuable for you. It might not make sense to add them *everywhere* but for something like this repo, or js-ipfs-api, I think they work great and give a really clean way to develop. YMMV. And I would imagine for newcomers it's easier to check which commands to run from a Makefile as opposed to package.json.

As mentioned above, Windows is the black sheep in this case too, so let's make sure any repo doesn't *depend* on Makefiles (ie. it's not the primary location to define the development tooling), they're additional info.

dignifiedquire replied:
You can get the same thing, cross platform with

```json
"scripts": {
  "clean": "rimraf node_modules",
  "test": "mytestcommand"
}
```

and `npm run clean && npm i && npm test`. If you want to have it even shorter, just add an alias to scripts 

```json
"make": "npm run clean && npm install && npm test"
```

haadcode replied:
Hey, I removed the Makefile from this PR as to not keep us back from merging this. I separated it to a new PR in https://github.com/ipfs/go-ipfs-dep/pull/15. 

This was a good discussion and at least the link to it should be kept safe for future references :)

haadcode commented:
Ok. Everything addressed. Can I get a CR and LGTM if all good to merge?

dignifiedquire commented on src/bin.js:
```diff
@@ -2,4 +2,9 @@
+// Third param is the target architecture
+require('./')(process.argv[2], process.argv[3], process.argv[4])
+  .then(() => process.exit(0))
+  .catch((e) => process.exit(1))
```
Wouldn't it be good to print the error and not just exit?

haadcode replied:
The errors (all output) are in index.js / error(). They get output, see the tests.

haadcode replied:
But you're right, the actual download function doesn't need to handle the output and it can be moved to bin.js (cli). Makes more sense in terms of responsibilities. Will change that.

haadcode replied:
Changed.

dignifiedquire commented on package.json:
```diff
@@ -2,6 +2,9 @@
   "description": "Install the latest go-ipfs binary",
+  "go-ipfs": {
+    "version": "v0.4.4"
+  },
```
Needs engines set to something like this: https://github.com/ipfs/js-ipfs-api/blob/master/package.json#L51-L54  as it is now using things like arrow functions, which are not available earlier

haadcode replied:
Fixed.

haadcode replied:
You can use one or the other. If you don't specify the "go-ipfs.version" field, then the package version is used. This is to enable us to release a) multiple version of the same go-ipfs version (eg. 0.4.5-1) and b) enable us to release dev versions of go-ipfs.

dignifiedquire commented:
@haadcode two last things, other than it looks good to me if CI is happy

dignifiedquire commented:
LGTM

diasdavid commented on .travis.yml:
```diff
@@ -10,16 +10,12 @@ before_install:
   - npm install -g npm
 
 script:
-  - npm run lint
```
Why stop linting?

diasdavid commented on .travis.yml:
```diff
@@ -10,16 +10,12 @@ before_install:
   - sh -e /etc/init.d/xvfb start
 
-after_success:
-  - npm run coverage-publish
```
Why stop publishing coverage?

haadcode replied:
Same as for linting.

diasdavid commented on package.json:
```diff
@@ -2,11 +2,16 @@
   "scripts": {
     "install": "node src/bin.js",
-    "lint": "aegir-lint",
-    "coverage": "aegir-coverage",
```
?

haadcode replied:
See above.

diasdavid commented on src/check-support.js:
```diff
@@ -0,0 +1,62 @@
+module.exports = {
+  Versions: supportedVersions,
+  Platforms: supportedPlatforms,
+  Archs: supportedArchs,
```
Do these have to be uppercase?

haadcode replied:
I prefer the upper case notation here.

diasdavid commented on test/index.js:
```diff
@@ -1,20 +1,145 @@
+const rimraf = require('rimraf')
+const goenv = require('go-platform')
+const pkg = require('./../package.json')
+const Download = require('../src')
```
Is this a class? If not, can we use lowercase?

haadcode replied:
Nope, not a class, just a function. See couple of lines below in the test file why that is wrapped like that.

haadcode commented:
@diasdavid Thanks for the review. I thought I had updated README in this PR but obviously didn't, so good catch (it was dropped somewhere down the line of merging and rebasing). Added it to the PR.

haadcode commented:
@diasdavid as for npm-go-ipfs, it only uses go-ipfs-dep to figure out a path (see https://github.com/ipfs/npm-go-ipfs/blob/master/link-ipfs.js#L9), so this PR will not effect it.

diasdavid commented on README.md:
````diff
@@ -17,8 +17,63 @@ npm install go-ipfs --save
+```json
+"go-ipfs": {
+  "version": "v0.4.5-pre1"
+},
````
super sweeeeeeeeet! ❤️❤️❤️❤️❤️❤️

diasdavid commented on README.md:
````diff
@@ -17,8 +17,63 @@ npm install go-ipfs --save
+  "version": "v0.4.5-pre1",
+  "distUrl": "http://localhost:8080/ipfs/QmSoNtqW22htkg9mtHWNBvZLUEmqfq8su7957meS1iQfeL"
+},
+```
````
❤️❤️❤️❤️❤️❤️❤️

haadcode replied:
I'm sure you got it, but just to make clear, this (the distUrl param) is to make it possible to a) test unreleased dist sites (so that whoever publishes one can be sure it doesn't break the whole of JS' go-ipfs ecosystem) and b) can test and use this via IPFS gateway (which makes it *extremely* fast) or locally.

diasdavid replied:
Yeah, that is way I was so excited :D
